src/cycles/eulerian.jl

```diff
@@ -0,0 +1,110 @@
+# Checking for existence of Eulerian circuit and Eulerian trail
+
+function isConnected(
+    graph::AbstractGraph{U},
+    ) where U<:Integer
+
+    visited = zeros(UInt8, nv(graph))
+
+    for v in vertices(graph)
+        if length(outneighbors(graph, v)) == 0
+            visited[v] = 2
+        end
+    end
+
+    for v in vertices(graph)
+
+        if length(outneighbors(graph, v)) == 0
+            continue
+        end
+        S = Vector{U}([v])
+        visited[v] = 1
+        while !isempty(S)
+            u = S[end]
+            w = 0
+            for n in outneighbors(graph, u)
+                if visited[n] == 0
+                    w = n
+                    break
+                end
+            end
+            if w != 0
+                visited[w] = 1
+                push!(S, w)
+            else
+                visited[u] = 2
+                pop!(S)
+            end
+        end
+
+        break
+    end
+
+    for v in vertices(graph)
+        if(visited[v]==0)
+            return false
+        end
+    end
+
+    return true
+end
+
+"""
+    eulerian_trail(g)
+
+Return `true` if graph `g` contains [Eulerian trail](https://en.wikipedia.org/wiki/Eulerian_path).
+
+### Implementation Notes
+Uses single call to iterative version of DFS.
+"""
+function eulerian_trail(
+    graph::AbstractGraph{U},
```

Comment:
If I see this correctly, this function only works for undirected graphs, right? If so, then you should restrict this in the function signature. You can look at the function `kurskal_mst` on how to do this.

Furthermore, in LightGraphs we usually use `T` instead of `U` for the vertex type.

Reply:
On writing
```
unction has_eulerian_trail end
@traitfn function has_eulerian_trail(
    graph::AbstractGraph::(!IsDirected),
    )
```
It gives error on testing
```
Eulerian traits and circuits: Error During Test at /home/chaman/git/LightGraphs.jl/test/cycles/eulerian.jl:10
  Test threw exception
  Expression: #= /home/chaman/git/LightGraphs.jl/test/cycles/eulerian.jl:10 =# @inferred has_eulerian_trail(g1)
  Not implemented: is_directed
  Stacktrace:
   [1] error(::String) at ./error.jl:33
   [2] _NI(::String) at /home/chaman/git/LightGraphs.jl/src/interface.jl:4
   [3] is_directed(::Type{AbstractGraph}) at /home/chaman/git/LightGraphs.jl/src/interface.jl:214
   [4] trait(::Type{IsDirected{AbstractGraph}}) at /home/chaman/.julia/packages/SimpleTraits/CZOA2/src/SimpleTraits.jl:172
   [5] has_eulerian_trail(::SimpleGraph{Int64}) at /home/chaman/.julia/packages/SimpleTraits/CZOA2/src/SimpleTraits.jl:350
   [6] macro expansion at /home/chaman/git/LightGraphs.jl/test/cycles/eulerian.jl:10 [inlined]
   [7] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1083 [inlined]
   [8] top-level scope at /home/chaman/git/LightGraphs.jl/test/cycles/eulerian.jl:4

```